File: ios/PSMobile/Screens/AppSettingsView.swift
import SwiftUI
import PSMShared

/// Einstellungen der App - Gegenstueck zu `AppSettingsScreen.kt`.
///
/// Bewusst getrennt von den Druckeinstellungen: dort geht es um das
/// Werkstueck, hier um das Programm. Wer die Modellvorschau abschalten
/// will, sucht sie nicht zwischen Schichthoehe und Fuelldichte.
///
/// Welche Schalter es gibt, in welcher Gruppe sie stehen und warum -
/// das alles kommt aus `AppSettings` im gemeinsamen Modul. Hier steht
/// nur, wie es auf iOS aussieht und wo die Werte liegen.
struct AppSettingsView: View {

    @ObservedObject var einstellungen: AppSettingsStore
    var onClose: () -> Void

    @Environment(\.psScale) private var ps
    @State private var zeigeSelbsttest = false
    @State private var protokollDatei: URL = LogExport.exportFile()
        ?? FileManager.default.temporaryDirectory.appendingPathComponent("psmobile-protokoll.txt")

    var body: some View {
        SchwebenderDialog(kennung: "dialog.appeinstellungen", maximaleBreite: ps.pt(760)) {
            VStack(spacing: 0) {
                kopfzeile
                Divider().background(PrusaColors.divider)
                ScrollView {
                    VStack(alignment: .leading, spacing: 0) {
                        ForEach(AppSettings.shared.groupsInOrder, id: \.self) { gruppe in
                            abschnitt(gruppe)
                        }
                        diagnose
                    }
                    .padding(.horizontal, ps.pt(16))
                    .padding(.bottom, ps.pt(24))
                    .frame(maxWidth: ps.pt(760))
                    .frame(maxWidth: .infinity)
                }
            }
        }
        .overlay(alignment: .topLeading) { PSMarke(name: "appeinstellungen") }
    }

    private var kopfzeile: some View {
        HStack(spacing: ps.pt(16)) {
            Button(action: onClose) {
                Text("‹  " + st("Back", "Zurück"))
                    .font(.system(size: ps.font(15)))
                    .foregroundStyle(PrusaColors.orange)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("appeinstellungen.zurueck")

            Text(st("App settings", "App-Einstellungen"))
                .font(.system(size: ps.font(20)))
                .foregroundStyle(PrusaColors.textPrimary)
            Spacer()
        }
        .padding(.horizontal, ps.pt(16))
        .frame(height: ps.touch(56))
    }

    @ViewBuilder private func abschnitt(_ gruppe: AppSettings.Group) -> some View {
        let schalter = AppSettings.shared.group(g: gruppe)
        let istDarstellung = gruppe == AppSettings.Group.appearance
        if !schalter.isEmpty || istDarstellung {
            Text(AppSettings.shared.groupTitle(g: gruppe).uppercased())
                .font(.system(size: ps.font(11), weight: .semibold))
                .foregroundStyle(PrusaColors.textMuted)
                .padding(.top, ps.pt(20))
                .padding(.bottom, ps.pt(6))

            // Sprache und Startmodus sind keine Ja/Nein-Fragen und stehen
            // deshalb als Auswahl in der Darstellung.
            if istDarstellung {
                auswahlZeile(
                    titel: st("Language", "Sprache"),
                    warum: st("Labels come from PrusaSlicer's own catalog.",
                              "Die Beschriftungen stammen aus PrusaSlicers eigenem Katalog."),
                    werte: ["en", "de"],
                    gewaehlt: einstellungen.language,
                    beschriftung: { $0 == "de" ? "Deutsch" : "English" },
                    kennung: "appeinstellungen.sprache",
                    waehlen: { einstellungen.language = $0 })
                auswahlZeile(
                    titel: st("On start", "Beim Start"),
                    warum: st("Skip the mode question if you always use the same one.",
                              "Die Modusfrage entfällt, wenn man ohnehin immer denselben nimmt."),
                    werte: AppSettings.shared.startModes,
                    gewaehlt: einstellungen.startMode,
                    beschriftung: { AppSettings.shared.startModeLabel(value: $0) },
                    kennung: "appeinstellungen.startmodus",
                    waehlen: { einstellungen.startMode = $0 })
            }

            ForEach(schalter, id: \.key) { eintrag in
                schalterZeile(eintrag)
            }
        }
    }

    /// Der Selbsttest, ganz unten.
    ///
    /// Er gehoert nicht zwischen die Schalter: das sind Vorlieben, das
    /// hier ist ein Werkzeug. Und er gehoert in die App und nicht in
    /// einen Testlauf am Rechner - was auf diesem Geraet gilt, weiss
    /// nur dieses Geraet.
    private var diagnose: some View {
        VStack(alignment: .leading, spacing: ps.pt(8)) {
            Text(st("Diagnostics", "Diagnose").uppercased())
                .font(.system(size: ps.font(11), weight: .semibold))
                .foregroundStyle(PrusaColors.textMuted)
                .padding(.top, ps.pt(20))
                .padding(.bottom, ps.pt(6))

            Button { zeigeSelbsttest = true } label: {
                HStack(spacing: ps.pt(12)) {
                    VStack(alignment: .leading, spacing: ps.pt(2)) {
                        Text(st("Self-test", "Selbsttest"))
                            .font(.system(size: ps.font(14)))
                            .foregroundStyle(PrusaColors.textPrimary)
                        Text(st("Runs loading, slicing, saving, painting and a load test on this device, and writes a report.",
                                "Prüft Laden, Schneiden, Sichern, Bemalen und eine Volllast auf diesem Gerät und schreibt einen Bericht."))
                            .font(.system(size: ps.font(11)))
                            .foregroundStyle(PrusaColors.textMuted)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                    Spacer()
                    Text("›")
                        .font(.system(size: ps.font(18)))
                        .foregroundStyle(PrusaColors.orange)
                }
                .padding(.horizontal, ps.pt(14))
                .padding(.vertical, ps.pt(10))
                .frame(minHeight: ps.touch(56))
                .background(PrusaColors.panelRaised)
                .clipShape(RoundedRectangle(cornerRadius: ps.pt(10)))
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("appeinstellungen.selbsttest")
            .sheet(isPresented: $zeigeSelbsttest) {
                SelbsttestView { zeigeSelbsttest = false }
            }

            ShareLink(item: protokollDatei) {
                HStack(spacing: ps.pt(12)) {
                    VStack(alignment: .leading, spacing: ps.pt(2)) {
                        Text(st("Export log", "Protokoll exportieren"))
                            .font(.system(size: ps.font(14)))
                            .foregroundStyle(PrusaColors.textPrimary)
                        Text(st("The last warnings and errors from the core, as a text file to share.",
                                "Die letzten Warnungen und Fehler aus dem Kern, als teilbare Textdatei."))
                            .font(.system(size: ps.font(11)))
                            .foregroundStyle(PrusaColors.textMuted)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                    Spacer()
                    Image(systemName: "square.and.arrow.up")
                        .font(.system(size: ps.font(16)))
                        .foregroundStyle(PrusaColors.orange)
                }
                .padding(.horizontal, ps.pt(14))
                .padding(.vertical, ps.pt(10))
                .frame(minHeight: ps.touch(56))
                .background(PrusaColors.panelRaised)
                .clipShape(RoundedRectangle(cornerRadius: ps.pt(10)))
                .contentShape(Rectangle())
            }
            // Bei jedem Erscheinen frisch schreiben statt einmal beim
            // ersten Aufbau der Ansicht - sonst zeigt der Export den
            // Stand von vor dem Oeffnen der Einstellungen, nicht den
            // aktuellen.
            .onAppear { protokollDatei = LogExport.exportFile() ?? protokollDatei }
            .accessibilityIdentifier("appeinstellungen.protokoll")

            versionszeile
        }
    }

    private var versionszeile: some View {
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "?"
        let build = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "?"
        return Text("PSMobile \(version) (\(build))")
            .font(.system(size: ps.font(11)))
            .foregroundStyle(PrusaColors.textMuted)
            .padding(.top, ps.pt(16))
            .frame(maxWidth: .infinity, alignment: .center)
            .accessibilityIdentifier("appeinstellungen.version")
    }

    private func schalterZeile(_ schalter: AppSettings.Toggle) -> some View {
        HStack(spacing: ps.pt(12)) {
            VStack(alignment: .leading, spacing: ps.pt(2)) {
                Text(schalter.title.text)
                    .font(.system(size: ps.font(14)))
                    .foregroundStyle(PrusaColors.textPrimary)
                Text(schalter.why.text)
                    .font(.system(size: ps.font(11)))
                    .foregroundStyle(PrusaColors.textMuted)
                    .fixedSize(horizontal: false, vertical: true)
            }
            Spacer()
            Toggle("", isOn: Binding(
                get: { einstellungen.bool(schalter.key, standard: schalter.standard) },
                set: { einstellungen.set(schalter.key, $0) }
            ))
            .labelsHidden()
            .tint(PrusaColors.orange)
            .accessibilityIdentifier("appeinstellungen.\(schalter.key)")
        }
        .padding(.horizontal, ps.pt(14))
        .padding(.vertical, ps.pt(10))
        .frame(minHeight: ps.touch(56))
        .background(PrusaColors.panelRaised)
        .clipShape(RoundedRectangle(cornerRadius: ps.pt(10)))
        .padding(.bottom, ps.pt(6))
    }

    private func auswahlZeile(titel: String,
                              warum: String,
                              werte: [String],
                              gewaehlt: String,
                              beschriftung: @escaping (String) -> String,
                              kennung: String,
                              waehlen: @escaping (String) -> Void) -> some View {
        VStack(alignment: .leading, spacing: ps.pt(8)) {
            VStack(alignment: .leading, spacing: ps.pt(2)) {
                Text(titel)
                    .font(.system(size: ps.font(14)))
                    .foregroundStyle(PrusaColors.textPrimary)
                Text(warum)
                    .font(.system(size: ps.font(11)))
                    .foregroundStyle(PrusaColors.textMuted)
                    .fixedSize(horizontal: false, vertical: true)
            }
            // Ein Menue statt einer Reihe Knoepfe: die Beschriftungen des
            // Startmodus sind zu lang fuer nebeneinander, und auf einem
            // iPhone im Hochformat bricht die Reihe sonst um.
            Menu {
                ForEach(werte, id: \.self) { wert in
                    Button(beschriftung(wert)) { waehlen(wert) }
                }
            } label: {
                HStack {
                    Text(beschriftung(gewaehlt))
                        .font(.system(size: ps.font(13)))
                        .foregroundStyle(PrusaColors.textPrimary)
                    Spacer()
                    Image(systemName: "chevron.up.chevron.down")
                        .font(.system(size: ps.font(10)))
                        .foregroundStyle(PrusaColors.textMuted)
                }
                .padding(.horizontal, ps.pt(12))
                .frame(height: ps.touch(44))
                .background(PrusaColors.background)
                .overlay(
                    RoundedRectangle(cornerRadius: ps.pt(6))
                        .stroke(PrusaColors.divider, lineWidth: 1)
                )
            }
            .accessibilityIdentifier(kennung)
        }
        .padding(.horizontal, ps.pt(14))
        .padding(.vertical, ps.pt(10))
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(PrusaColors.panelRaised)
        .clipShape(RoundedRectangle(cornerRadius: ps.pt(10)))
        .padding(.bottom, ps.pt(6))
    }

    private func st(_ english: String, _ german: String) -> String {
        SimpleModeState.shared.text(english: english, german: german)
    }
}

/// Die Werte hinter den App-Einstellungen.
///
/// Gegenstueck zu den SharedPreferences auf Android: dieselben
/// Schluessel aus `AppSettings`, nur in `UserDefaults`. Die Schluessel
/// stehen im gemeinsamen Modul, damit eine spaeter ergaenzte Einstellung
/// nicht auf einer Seite vergessen wird.
@MainActor
final class AppSettingsStore: ObservableObject {

    @Published var language: String {
        didSet {
            UserDefaults.standard.set(language, forKey: Self.languageKey)
            // Der Katalog haelt sowohl die Beschriftungen als auch
            // Lang.current fuer das gemeinsame Modul.
            PsUiCatalog.load(language: language)
        }
    }

    @Published var startMode: String {
        didSet { UserDefaults.standard.set(startMode, forKey: AppSettings.shared.KEY_START_MODE) }
    }

    /// Nur damit die Oberflaeche sich neu zeichnet - die Wahrheit steht
    /// in UserDefaults.
    @Published private var revision = 0

    private static let languageKey = "ui.language"

    init() {
        language = UserDefaults.standard.string(forKey: Self.languageKey) ?? "en"
        startMode = UserDefaults.standard.string(forKey: AppSettings.shared.KEY_START_MODE)
            ?? AppSettings.shared.START_ASK
    }

    func bool(_ key: String, standard: Bool) -> Bool {
        _ = revision
        // `object(forKey:)` unterscheidet "nicht gesetzt" von "false" -
        // `bool(forKey:)` kann das nicht, und ein Schalter mit Standard
        // "an" waere beim ersten Start faelschlich aus.
        guard UserDefaults.standard.object(forKey: key) != nil else { return standard }
        return UserDefaults.standard.bool(forKey: key)
    }

    func set(_ key: String, _ value: Bool) {
        UserDefaults.standard.set(value, forKey: key)
        revision += 1
    }

    var thumbnails: Bool {
        bool(AppSettings.shared.KEY_THUMBNAILS, standard: true)
    }

    var showIncompatible: Bool {
        bool(AppSettings.shared.KEY_SHOW_INCOMPATIBLE, standard: false)
    }

    var multiBedRender: Bool {
        bool(AppSettings.shared.KEY_MULTI_BED_RENDER, standard: true)
    }
}

File: android/app/src/main/java/de/psmobile/ui/AppSettingsView.kt
package de.psmobile.ui

// `AppSettingsView.swift` traegt drueben zwei Dinge: den Bildschirm und
// die Klasse `AppSettingsStore`. Auf Android steht der Store in einer
// eigenen Datei, `AppSettingsStore.kt` - deshalb landen Aenderungen am
// Store dort und nicht hier.
//
// 10.09.2026: `-psm-reset-settings` setzt die App-Einstellungen auf die
// Voreinstellung aus dem gemeinsamen Modul zurueck. Der Bildvergleich
// verglich sonst zwei Vorgeschichten statt zweier Oberflaechen. Die
// Android-Fassung steht im `init`-Block von `AppSettingsStore.kt`.
// 12.09.2026: dazu gehoert jetzt auch `psm.sichtbarkeit`, die Stufe
// Simple/Advanced/Expert der Einstellungsseiten - ebenfalls dort.

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import de.psmobile.BuildConfig
import de.psmobile.core.LogExport
import de.psmobile.shared.rules.AppSettings
import de.psmobile.ui.theme.PrusaColors
import de.psmobile.ui.theme.ScaledOverlay

/// Einstellungen der App.
///
/// Bewusst getrennt von den Druckeinstellungen: dort geht es um das
/// Werkstueck, hier um das Programm. Welche Schalter es gibt, in
/// welcher Gruppe sie stehen und warum - das alles kommt aus
/// `AppSettings` im gemeinsamen Modul.
@Composable
fun AppSettingsView(
    einstellungen: AppSettingsStore,
    onClose: () -> Unit,
) {
    val ps = LocalPsScale.current
    var zeigeSelbsttest by remember { mutableStateOf(false) }
    var zeigeOpenSource by remember { mutableStateOf(false) }

    BackHandler(onBack = onClose)

    Box {
        SchwebenderDialog(kennung = "dialog.appeinstellungen", maximaleBreite = ps.pt(760)) {
            Column {
                Kopfzeile(onClose)
                HorizontalDivider(color = PrusaColors.divider)
                Column(
                    Modifier
                        // Die Liste traegt einen Namen, damit ein Test zur
                        // Diagnose ganz unten blaettern kann - siehe die
                        // gleichlautende Stelle in AppSettingsView.swift.
                        .testTag("appeinstellungen.liste")
                        .verticalScroll(rememberScrollState())
                        .fillMaxWidth()
                        .padding(horizontal = ps.pt(16))
                        .padding(bottom = ps.pt(24)),
                    horizontalAlignment = Alignment.Start,
                ) {
                    AppSettings.groupsInOrder.forEach { gruppe ->
                        Abschnitt(gruppe, einstellungen)
                    }
                    Diagnose(onSelbsttest = { zeigeSelbsttest = true })
                    Ueber(onOpenSource = { zeigeOpenSource = true })
                }
            }
        }
        PSMarke(name = "appeinstellungen")
    }

    if (zeigeOpenSource) {
        Dialog(
            onDismissRequest = { zeigeOpenSource = false },
            properties = DialogProperties(usePlatformDefaultWidth = false),
        ) {
            ScaledOverlay {
                Box(Modifier.fillMaxSize()) {
                    OpenSourceView(onClose = { zeigeOpenSource = false })
                }
            }
        }
    }

    if (zeigeSelbsttest) {
        Dialog(
            onDismissRequest = { zeigeSelbsttest = false },
            properties = DialogProperties(usePlatformDefaultWidth = false),
        ) {
            ScaledOverlay {
                Box(Modifier.fillMaxSize()) {
                    SelbsttestView(onClose = { zeigeSelbsttest = false })
                }
            }
        }
    }
}

@Composable
private fun Kopfzeile(onClose: () -> Unit) {
    val ps = LocalPsScale.current
    Row(
        Modifier
            .fillMaxWidth()
            .height(ps.touch(56))
            .padding(horizontal = ps.pt(16)),
        horizontalArrangement = Arrangement.spacedBy(ps.pt(16)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            "‹  " + st("Back", "Zurück"),
            Modifier
                .clickable(onClick = onClose)
                .testTag("appeinstellungen.zurueck"),
            fontSize = ps.font(15), color = PrusaColors.orange,
        )
        Text(st("App settings", "App-Einstellungen"), fontSize = ps.font(20), color = PrusaColors.textPrimary)
        Spacer(Modifier.weight(1f))
    }
}

@Composable
private fun Abschnitt(gruppe: AppSettings.Group, einstellungen: AppSettingsStore) {
    val ps = LocalPsScale.current
    val schalter = AppSettings.group(gruppe)
    val istDarstellung = gruppe == AppSettings.Group.APPEARANCE
    if (schalter.isNotEmpty() || istDarstellung) {
        Text(
            AppSettings.groupTitle(gruppe).uppercase(),
            Modifier.padding(top = ps.pt(20), bottom = ps.pt(6)),
            fontSize = ps.font(11), fontWeight = FontWeight.SemiBold, color = PrusaColors.textMuted,
        )

        // Sprache und Startmodus sind keine Ja/Nein-Fragen und stehen
        // deshalb als Auswahl in der Darstellung.
        if (istDarstellung) {
            AuswahlZeile(
                titel = st("Language", "Sprache"),
                warum = st(
                    "Labels come from PrusaSlicer's own catalog.",
                    "Die Beschriftungen stammen aus PrusaSlicers eigenem Katalog.",
                ),
                werte = listOf("en", "de"),
                gewaehlt = einstellungen.language,
                beschriftung = { if (it == "de") "Deutsch" else "English" },
                kennung = "appeinstellungen.sprache",
                waehlen = { einstellungen.language = it },
            )
            AuswahlZeile(
                titel = st("On start", "Beim Start"),
                warum = st(
                    "Skip the mode question if you always use the same one.",
                    "Die Modusfrage entfällt, wenn man ohnehin immer denselben nimmt.",
                ),
                werte = AppSettings.startModes,
                gewaehlt = einstellungen.startMode,
                beschriftung = { AppSettings.startModeLabel(it) },
                kennung = "appeinstellungen.startmodus",
                waehlen = { einstellungen.startMode = it },
            )
        }

        schalter.forEach { eintrag ->
            SchalterZeile(eintrag, einstellungen)
        }
    }
}

/// Der Selbsttest und der Protokoll-Export, ganz unten. Das sind
/// Werkzeuge, keine Vorlieben.
@Composable
private fun Diagnose(onSelbsttest: () -> Unit) {
    val ps = LocalPsScale.current
    val context = LocalContext.current
    Column(verticalArrangement = Arrangement.spacedBy(ps.pt(8)), horizontalAlignment = Alignment.Start) {
        Text(
            st("Diagnostics", "Diagnose").uppercase(),
            Modifier.padding(top = ps.pt(20), bottom = ps.pt(6)),
            fontSize = ps.font(11), fontWeight = FontWeight.SemiBold, color = PrusaColors.textMuted,
        )

        Row(
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(ps.pt(10)))
                .background(PrusaColors.panelRaised)
                .clickable(onClick = onSelbsttest)
                .testTag("appeinstellungen.selbsttest")
                .heightIn(min = ps.touch(56))
                .padding(horizontal = ps.pt(14), vertical = ps.pt(10)),
            horizontalArrangement = Arrangement.spacedBy(ps.pt(12)),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(ps.pt(2))) {
                Text(st("Self-test", "Selbsttest"), fontSize = ps.font(14), color = PrusaColors.textPrimary)
                Text(
                    st(
                        "Runs loading, slicing, saving, painting and a load test on this device, and writes a report.",
                        "Prüft Laden, Schneiden, Sichern, Bemalen und eine Volllast auf diesem Gerät und schreibt einen Bericht.",
                    ),
                    fontSize = ps.font(11), color = PrusaColors.textMuted,
                )
            }
            Text("›", fontSize = ps.font(18), color = PrusaColors.orange)
        }

        // Frisch geschrieben beim Antippen, nicht einmal beim ersten
        // Aufbau - sonst zeigt der Export einen alten Stand.
        Row(
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(ps.pt(10)))
                .background(PrusaColors.panelRaised)
                .clickable {
                    val uri = runCatching { LogExport.uri(context) }.getOrNull() ?: return@clickable
                    teilen(context, uri, "text/plain", "psmobile-protokoll.txt")
                }
                .testTag("appeinstellungen.protokoll")
                .heightIn(min = ps.touch(56))
                .padding(horizontal = ps.pt(14), vertical = ps.pt(10)),
            horizontalArrangement = Arrangement.spacedBy(ps.pt(12)),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(ps.pt(2))) {
                Text(st("Export log", "Protokoll exportieren"), fontSize = ps.font(14), color = PrusaColors.textPrimary)
                Text(
                    st(
                        "The last warnings and errors from the core, as a text file to share.",
                        "Die letzten Warnungen und Fehler aus dem Kern, als teilbare Textdatei.",
                    ),
                    fontSize = ps.font(11), color = PrusaColors.textMuted,
                )
            }
            SfSymbol("square.and.arrow.up", Modifier.size(ps.pt(16)), tint = PrusaColors.orange)
        }

    }
}

/// Woraus die App besteht und wo ihr Quellcode liegt - PrusaSlicers
/// AGPL verlangt beides sichtbar in der App. Ganz unten, unter der
/// Diagnose, mit der Versionszeile.
@Composable
private fun Ueber(onOpenSource: () -> Unit) {
    val ps = LocalPsScale.current
    Column(verticalArrangement = Arrangement.spacedBy(ps.pt(8)), horizontalAlignment = Alignment.Start) {
        Text(
            st("About", "Über").uppercase(),
            Modifier.padding(top = ps.pt(20), bottom = ps.pt(6)),
            fontSize = ps.font(11), fontWeight = FontWeight.SemiBold, color = PrusaColors.textMuted,
        )
        Row(
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(ps.pt(10)))
                .background(PrusaColors.panelRaised)
                .clickable(onClick = onOpenSource)
                .testTag("appeinstellungen.opensource")
                .heightIn(min = ps.touch(56))
                .padding(horizontal = ps.pt(14), vertical = ps.pt(10)),
            horizontalArrangement = Arrangement.spacedBy(ps.pt(12)),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(ps.pt(2))) {
                Text(st("Open source & licenses", "Open Source & Lizenzen"), fontSize = ps.font(14), color = PrusaColors.textPrimary)
                Text(
                    st(
                        "Based on PrusaSlicer (AGPL-3.0). Source code on GitHub and the list of third-party components.",
                        "Basiert auf PrusaSlicer (AGPL-3.0). Quellcode auf GitHub und die Liste der verwendeten Komponenten.",
                    ),
                    fontSize = ps.font(11), color = PrusaColors.textMuted,
                )
            }
            Text("›", fontSize = ps.font(18), color = PrusaColors.orange)
        }
        Versionszeile()
    }
}

@Composable
private fun Versionszeile() {
    val ps = LocalPsScale.current
    Text(
        "Slicer Mobile ${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
        Modifier
            .padding(top = ps.pt(16))
            .fillMaxWidth()
            .testTag("appeinstellungen.version"),
        fontSize = ps.font(11), color = PrusaColors.textMuted, textAlign = TextAlign.Center,
    )
}

@Composable
private fun SchalterZeile(schalter: AppSettings.Toggle, einstellungen: AppSettingsStore) {
    val ps = LocalPsScale.current
    Row(
        Modifier
            .fillMaxWidth()
            .padding(bottom = ps.pt(6))
            .clip(RoundedCornerShape(ps.pt(10)))
            .background(PrusaColors.panelRaised)
            .heightIn(min = ps.touch(56))
            .padding(horizontal = ps.pt(14), vertical = ps.pt(10)),
        horizontalArrangement = Arrangement.spacedBy(ps.pt(12)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(ps.pt(2))) {
            Text(schalter.title.text, fontSize = ps.font(14), color = PrusaColors.textPrimary)
            Text(schalter.why.text, fontSize = ps.font(11), color = PrusaColors.textMuted)
        }
        Switch(
            checked = einstellungen.bool(schalter.key, schalter.standard),
            onCheckedChange = { einstellungen.set(schalter.key, it) },
            modifier = Modifier.testTag("appeinstellungen.${schalter.key}"),
            colors = SwitchDefaults.colors(
                checkedThumbColor = Color.White,
                checkedTrackColor = PrusaColors.orange,
            ),
        )
    }
}

@Composable
private fun AuswahlZeile(
    titel: String,
    warum: String,
    werte: List<String>,
    gewaehlt: String,
    beschriftung: (String) -> String,
    kennung: String,
    waehlen: (String) -> Unit,
) {
    val ps = LocalPsScale.current
    var offen by remember { mutableStateOf(false) }
    Column(
        Modifier
            .fillMaxWidth()
            .padding(bottom = ps.pt(6))
            .clip(RoundedCornerShape(ps.pt(10)))
            .background(PrusaColors.panelRaised)
            .padding(horizontal = ps.pt(14), vertical = ps.pt(10)),
        verticalArrangement = Arrangement.spacedBy(ps.pt(8)),
        horizontalAlignment = Alignment.Start,
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(ps.pt(2)), horizontalAlignment = Alignment.Start) {
            Text(titel, fontSize = ps.font(14), color = PrusaColors.textPrimary)
            Text(warum, fontSize = ps.font(11), color = PrusaColors.textMuted)
        }
        // Ein Menue statt einer Reihe Knoepfe: die Beschriftungen des
        // Startmodus sind zu lang fuer nebeneinander.
        Box {
            val form = RoundedCornerShape(ps.pt(6))
            Row(
                Modifier
                    .fillMaxWidth()
                    .height(ps.touch(44))
                    .background(PrusaColors.background, form)
                    .border(1.dp, PrusaColors.divider, form)
                    .clickable { offen = true }
                    .testTag(kennung)
                    .padding(horizontal = ps.pt(12)),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(beschriftung(gewaehlt), fontSize = ps.font(13), color = PrusaColors.textPrimary)
                Spacer(Modifier.weight(1f))
                SfSymbol("chevron.up.chevron.down", Modifier.size(ps.pt(10)), tint = PrusaColors.textMuted)
            }
            DropdownMenu(expanded = offen, onDismissRequest = { offen = false }) {
                ScaledOverlay {
                    Column {
                        werte.forEach { wert ->
                            DropdownMenuItem(
                                text = { Text(beschriftung(wert)) },
                                onClick = {
                                    offen = false
                                    waehlen(wert)
                                },
                            )
                        }
                    }
                }
            }
        }
    }
}
